frame/support/procedural/src/construct_runtime/mod.rs

```diff
@@ -425,3 +442,38 @@ fn decl_integrity_test(scrate: &TokenStream2) -> TokenStream2 {
 		}
 	)
 }
+
+/// Assign index to each modules using same rules as rust for fieldless enum.
+/// I.e. implicit are assigned number incrementedly from last explicit or 0.
+fn assign_implicit_index(modules: &mut Vec<ModuleDeclaration>) -> syn::Result<()> {
+	let mut indices = HashMap::new();
+	let mut last_index: Option<u8> = None;
+
+	for module in modules {
+		match module.index {
+			Some(index) => {
+				last_index = Some(index);
+			},
+			None => {
+				let index = match last_index.map_or(Some(0), |i| i.checked_add(1)) {
+					Some(i) => i,
+					None => {
+						let msg = "module index doesn't fit into u8, index is 256";
+						return Err(syn::Error::new(module.module.span(), msg));
+					},
+				};
+				module.index = Some(index);
+				last_index = Some(index);
+			},
+		}
+
+		if let Some(used_span) = indices.insert(module.index.unwrap(), module.module.span()) {
+			let span = module.module.span().join(used_span)
+				.expect("Modules are defined in same file");
+			let msg = format!("module index are conflicting, at index {}", module.index.unwrap());
```

Comment:
If you return the index above from the match. You don't need to use `unwrap` here. This could also simplify the match above.

Reply:
I factorized a bit